WIP : Repository Refactor/Revamp

troussej opened the pull request:
This PR is a work in progress, I opened	it to start a discussion before going further

My initial goal was to make a new page which would serve as a cross repository explorer, with visual editing features and full ajax queries to be quicker.

To achieve that, I needed to leverage part of the repo's page methods (display as tab, tree) and ended refactoring it quite a lot, not only the "backend part" but also the repo page itsef : display as tab, edit inline, link to child item.

I'm now at a state with a almost full iso-functionnal repository page, with a new "engine" under the hood, and also fixed some bugs / inconsistencies in the UI/UX.

In short, what's new:

- New "MVC" model that uses classes to formalize models:
    - Using objects makes it clearer, less arrays everywhere that store bits of data.
    - Repository :
        - holds the whole repo definition in memory for when using multiple repositories.
    - ItemDescriptor :
        - contains self and parent properties
        - xmlDef is parsed on demand and stored for further use
    - PropertyDescriptor : 
        - stores default value, item type of property if required.
    - RepositoryItem :  model of a parsed xml result
    - PropertyValue : each value, default or actual, with its flags parsed from the xml (rdonly, etc )
- Result as tab:
    - Fully compatible and agnostic between itemtree and standard result (+ repo explorer when it's done)
    - properties sorted alphabetically
    - load sub item :
        - now uses an ajax call, saves the time to reload the page & BDA - **several seconds**. 
        - If the item is already on the page, **scroll to it**. (like for item tree)
        - Now available for multi items (lists & maps)
    - more constistent collapse expand of long properties:
        - added collapse all button
        - per cell collapse/expand
    - inline edition more consistent across fields types (not finished yet)
    - TODO : reload item view, delete view
- TODO: 
    - change the item tree generation to use the same model for each type of visualization, instead of reloading the tree each time.
    - add help/hints overlays.
    - performance testing

Commit:
feat(dash): wip - reuse repo display using xmlDef

Commit:
feat(dash): use repo display

Commit:
feat(dash): use repo page display and sub elems for print/query

Commit:
fix: getXmlDef

Commit:
init repoExplo tab

Commit:
init repo page

Commit:
fix broken js

Commit:
fix default tab on reopen

Commit:
feat: print request

Commit:
refactor inputs

Commit:
feat(repo-explorer): init autocomplete

Commit:
feat(repo-explorer): update item descriptor list on repo update

Commit:
wrap repo select2 style

Commit:
add current repo button

Commit:
fix:  trigger change event on repo manual change

Commit:
feat(search): use favs as local base for autocomplete + add to repo explorer

Commit:
feat: bind clear all button

Commit:
adjust dash modal size

Commit:
fix: display repo error in dash

Commit:
feat : add margin to help prompt in err msg

Commit:
fix: change dash completion priority

Commit:
bind enter + handle empty response

Commit:
fix fields init editable state

Commit:
build itemdescriptors recursively

Commit:
build repoItem

Commit:
format as table with new model

Commit:
render objects as table

Commit:
handle default value

Commit:
display derivation flags

Commit:
add manual ellipsis for long fields

Commit:
refactor result formating + plug to dash

Commit:
feat(repostory): load sub item by ajax call

Commit:
feat(repository): handle map of ids

Commit:
refactor(repo): align icons, refactor cell building

Commit:
feat(repo): scroll to item if already loaded

Commit:
feat(repo): add speedbar to new format

Commit:
refactor sub item link

Commit:
inline edit repo item

troussej commented:
also : I'm considering adding Q as a new library to handle Promises, there are a lot of nested callbacks, it has become quite unreadable sometimes.

jc7447 commented:
Outch, That's a lot of stuff !

I will need some time to digest since I don't recognize a lot the code anymore ;)

troussej commented:
Take your time. I don't need to merge right now, I can continue to work on the branch.

Commit:
fix(repo): xml tab : css align actions

Commit:
fix(repo): inline edit hide actions

Commit:
feat(repo): reload item

Commit:
fix: close tag

Commit:
fix: rql form invisible

Commit:
fix: update sucess notify not showing

Commit:
feat(repo): reload item after update

Commit:
fix scrollTo

Commit:
fix: inline edit in dash

Commit:
dash:make it real fullscreen

Commit:
feat(repo): copy xml from tab

Commit:
feat: close xml tab

troussej commented:
Added a lot of fixes.

Plus : reload an item, copy item xml from tab, better inline edit, close tab.

troussej commented:
I'm globally satisfied with the changes, so it's ready to merge when you are.

Commit:
feat(repo): replace inline forms inputs by textareas

Commit:
add quicklinks to standard results

Commit:
fix: focus inline edit

Commit:
sub property of other repository

Commit:
when load items, append to existing table

Commit:
fix : reload speedbar when closing tab

troussej commented:
Hi @jc7447 

Did you get a chance to look at it? I would like to merge this and work on some other stuff.
We also should discuss the next release date, there have been quite some changes since april.

Commit:
add a flash when loading/reloading a tab

Commit:
fix(repo): use result itemDescriptor to append elems

jc7447 commented:
Hi,

I'm time short those days but I will try to review this next week, once it's done I will release version 2.2

troussej commented:
No pressure ;)

Le 19 nov. 2017 11:42 AM, "Jean-Charles Manoury" <notifications@github.com>
a écrit :

> Hi,
>
> I'm time short those days but I will try to review this next week, once
> it's done I will release version 2.2
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jc7447/BetterDynAdmin/pull/132#issuecomment-345507287>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH7cnK7PUbuK02yAsz89D5JLgKRyUJodks5s4AYYgaJpZM4QMwoV>
> .
>

Commit:
highlight default values

Commit:
fix(repo): when appending to existing tab, show lines that are not empty

jc7447 reviewed (commented):
Finally take some time to look at this closer :)

I noticed a few things when testing :

 - It's quite slow when the number of items increase, for example when I printed 100 orders (probably due to the amount of log ) :

BDA 2.1 :
sanitizeXml: 2.08984375ms
renderTab: 270.8701171875ms
bdaRepository: 631.1357421875ms
bda: 674.686279296875ms

Repo_explorer :
showXMLAsTab_new: 19490.670166015625ms
preparePage: 19568.89794921875ms
showQueryList: 11.4130859375ms
setupItemTreeForm: 30.7978515625ms
setupItemDescriptorTable: 233.130859375ms
setupPropertiesTables: 2.7021484375ms
setupRepositoryCacheSection: 186.809814453125ms
initCodeMirror: 46.473876953125ms
initSelect2: 69.615966796875ms
bdaRepository: 20219.025146484375ms

- It would be nice to have tooltips for the new button on top of each item (I thought that the cross was a remove-item action)

- I got an exception when trying to use the Repo Explorer in DASH, I was not able to see descriptors when choosing a repository :
Error: datumTokenizer is required
    at Function.error (eval at E_c (:3:298), <anonymous>:6:1971)
    at eval (eval at E_c (:3:298), <anonymous>:472:10643)
    at new c (eval at E_c (:3:298), <anonymous>:472:10941)
    at Object.initRepositoryField (eval at E_c (:3:298), <anonymous>:12483:32)
    at Object.initFields (eval at E_c (:3:298), <anonymous>:12424:22)
    at HTMLBodyElement.$.on (eval at E_c (:3:298), <anonymous>:12418:24)
    at HTMLBodyElement.dispatch (eval at E_c (:3:298), <anonymous>:7:9856)
    at HTMLBodyElement.q.handle (eval at E_c (:3:298), <anonymous>:7:7883)
    at Object.trigger (eval at E_c (:3:298), <anonymous>:8:5413)
    at HTMLBodyElement.eval (eval at E_c (:3:298), <anonymous>:8:5923)

- The "show speedbar" button does nothing (no error in the log)

- Read-only value are shown in grey like default value

- I got this exception in a ProductCatalog page : 
TypeError: Cannot read property 'properties' of undefined
    at _.each (eval at E_c (:3:298), <anonymous>:3663:25)
    at eval (eval at E_c (:3:298), <anonymous>:455:26034)
    at oe (eval at E_c (:3:298), <anonymous>:455:16451)
    at eval (eval at E_c (:3:298), <anonymous>:455:25844)
    at Function.gf (eval at E_c (:3:298), <anonymous>:456:9648)
    at Y.n.(anonymous function) [as each] (eval at E_c (:3:298), <anonymous>:456:22529)
    at Object.formatTabResult (eval at E_c (:3:298), <anonymous>:3660:12)
    at Object.renderResultSection (eval at E_c (:3:298), <anonymous>:3585:22)
    at Object.showXMLAsTab (eval at E_c (:3:298), <anonymous>:3537:22)
    at eval (eval at E_c (:3:298), <anonymous>:4410:34)

The XML of this repository was never correctly parsed by JQuery, so there was always an error somewhere but the result was displayed, now the page just breaks. I will provide you more detail privately.

- The anchor links in the item tree don't go exactly to the selected item, it's a bit confusing

troussej commented:
TODO
- [x] perfs
   - [x] create edition form on click, not on display (almost 30% of time spent here)


- [x] tooltip/help - I was thinking the same

- [x] dash - i'm going to remove the new tab for now, as the feature it not quite finished (the repo refactor took over)

- [x] speedbar button

- [x] read only vs default values

- [ ] exception handling

- [x] anchor links : i used a new method from html5 to scrollTo, it has weird behavior sometimes, depending on the browser. One good thing is that is also scrolls left/right. I'll check if there are more options to tune it, and also maybe flash the item that is scrolled to (it's confusing as hell when there are a lot of priceInfo for example).

Commit:
revert(reex): remove explorer for now

Commit:
add time traces for perfs debug

troussej commented:
just rebased from dev to get the latest refactors (some about perfs)

troussej commented:
perfs : looks like the culprit is buildResultTable, it is linear with the number of items, about 100ms / item

Commit:
comment out stuff to test perfs

Commit:
refactor:  move action bind globally

Commit:
refactor(repo): perfs - bind all actions at once

Commit:
refactor(repo): perfs - don't handle non visible lines

Commit:
feat(perfs): add new class PerformanceMonitor

Commit:
add more perfs logs

Commit:
add average to perfMon

Commit:
add log info and fix getXmlDef

Commit:
feat(repo): perfs - create edit form on demand

Commit:
feat(repo): build xml tab is asynchronous

Commit:
feat(perfs): add more monitoring

Commit:
feat(perfs): replace dom manipulation by string concat

Commit:
fix(repo): fix long line toggle

troussej commented:
I'm replacing all the dom manipulation by string concatenation, it's around 5 times faster.

Commit:
fix(perfs): create link to other items using strings

Commit:
fix: load savd query

Commit:
feat(repo): add tooltips

Commit:
feat(repository): change style of readonly properties

Commit:
feat(repository): add progress bar for item rendering

Commit:
fix(dash): init repo perfMonitor in build function

Commit:
fix: align scrollTo on top

Commit:
feat(repository: re-add speedbar

Commit:
fix(repository):  show speedbar button

Commit:
fix(repository): fix scrollTo

Commit:
fix: loading icon, multi speedbar button

Commit:
chore: cleanup - remove old code

Commit:
feat(repo): add bluebird library to use Promises

Commit:
feat(repo): wip - promisify rendering to allow progressive loading

Commit:
wip: progressive loading

Commit:
fix(repo): fix speedbar - use manager to throtle scroll events

 - for perfs just one scroll event
 - scroll event does not register after a promise, don't know why. With the manager, the scroll event is bound before doing anything
~

troussej commented:
Hi @jc7447

I finally got time to fix the last bugs.

Perfs are much better (even if slower than before, we are doing more calculations)
I tried to make it smoother for the user, by adding a loading bar and trying to defer some calculations
Added a "speedbarManager" to handle the scroll events, for perfs.

troussej committed:
Update bda.xmldef.js

do not collapse table def on click
